compiler/modules/bir/verify.bal

```diff
@@ -65,13 +65,223 @@ class VerifyContext {
+    }
+    if term is BranchInsn {
+        check verifyChildStructure(vc, blocks, term.dest, fwdInDegree, preds, termPos, term.backward);
+    }
+    else if term is TypeBranchInsn|CondBranchInsn {
+        check verifyChildStructure(vc, blocks, term.ifTrue, fwdInDegree, preds, termPos);
+        check verifyChildStructure(vc, blocks, term.ifFalse, fwdInDegree, preds, termPos);
+    }
+    preds[current] = false;
+    fwdInDegree[current] = 1; // fwdInDegree[current] was 0 until now since above recursions can't cycle back to self.
+}
+
+function verifyChildStructure(VerifyContext vc, BasicBlock[] blocks, Label child, int[] fwdInDegree, boolean[] preds, Position pos, boolean backward = false) returns Error? {
+    boolean cycle = preds[child];
+    if backward != cycle {
+        return vc.invalidErr(backward ? "backward branch to non-predecessor" : "forward branch form a cycle", pos);
+    }
+    if backward {
+        return;
+    }
+    int fid = fwdInDegree[child];
+    if fid == 0 { // unvisited
+        check verifyStructure(vc, blocks, child, fwdInDegree, preds, pos);
+    }
+    else {
+        fwdInDegree[child] = fid + 1;
+    }
+}
+
+function verifyReachable(VerifyContext vc, int[] fwdInDegree, Position pos) returns Error? {
+    foreach int fid in fwdInDegree {
+        if fid == 0 {
+            return vc.invalidErr("block is not reachable", pos);
+        }
+    }
+}
+
+// Verify param registers are at the beginning of the array. Returns a boolean array marking params as initialized (true).
+function verifyParamRegs(VerifyContext vc, Register[] regs) returns boolean[]|Error {
+    boolean[] paramRegs = [];
+    boolean afterParams = false;
+    foreach var reg in regs {
+        if reg is ParamRegister {
+            if afterParams {
+                return vc.invalidErr("param register is not at the beginning of the register list", reg.pos);
+            }
+            paramRegs.push(true);
+        }
+        else {
+            paramRegs.push(false);
+            afterParams = true;
+        }
+    }
+    return paramRegs;
+}
+
+// Verify Registers are initialized before used.
+function verifyRegFlow(VerifyContext vc, BasicBlock[] blocks, Label current, int[] fwdInDegrees, RegFlow[][] blocksFlows, int? pred, boolean[] predRegs, Position predPos) returns Error? {
+    BasicBlock block = blocks[current];
+    RegFlow[] flows = blocksFlows[current];
+    flows.push({ origin: pred, regs: predRegs });
+    if fwdInDegrees[current] != flows.length() {
+        return;
+    }
+    boolean[] regs = mergeFlows(flows);
+    boolean afterMerge = false;
+    foreach var insn in block.insns {
+        if insn is TypeMergeInsn {
+            check verifyTypeMergeFlow(vc, afterMerge, flows, insn);
+        }
+        else {
+            afterMerge = true;
+        }
+        match insn {
+            var { result } => {
+                regs[result.number] = true;
```

Comment:
I need to special case ListSetInsn.
1. Why isn't this failing ?
2. May I go back to how ListSetInsn originally was. ie: having a separate field for result.

Reply:
ListSetInsn is not a special case. It only has operands no result.